docs/blog/posts/control-flow-analysis.md

```diff
@@ -0,0 +1,115 @@
+## Showing that all valid messages are committed
+The property we showed in the previous section is a safety property: we show that we cannot end in a bad state by committing a message that has not been validated. But this doesn't mean we actually commit any message. Let's modify our specification to show that we do commit messages:
+```rust
+#[hax_lib::fstar::before(r"
+assume val message_validated: string -> Type0
+assume val message_committed: string -> Type0
+")]
+#[hax_lib::opaque]
+#[hax_lib::ensures(|_| fstar!("message_validated $msg"))]
+fn validate_message(msg: &str) -> Result<(), Error> {
+    if msg == "" {
+        Err(Error::InvalidMessage)
+    } else {
+        Ok(())
+    }
+}
+
+#[hax_lib::attributes]
+impl State {
+    #[hax_lib::requires(fstar!("message_validated $msg"))]
+    #[hax_lib::opaque]
+    #[hax_lib::ensures(|_| fstar!("message_committed $msg"))]
+    fn commit_message(&mut self, msg: String) {
+        self.0.push(msg)
+    }
+
+    #[hax_lib::ensures(|res| fstar!("message_committed $msg"))]
+    fn receive_message(&mut self, msg: String)-> Result<(), Error> {
+        validate_message(&msg)?;
+        self.commit_message(msg);
+        Ok(())
+    }
+}
+```
+We added a new abstract predicate `message_committed` that holds if a message has been committed. With this specification, F* is not able to prove the post-condition for `receive_message`. Of course, the message is committed only when it is valid so we need to add this condition in our post-condition:
+```rust
+#[hax_lib::ensures(|res| hax_lib::implies(validate_message(&msg).is_ok(), fstar!("message_committed $msg")))]
+fn receive_message(&mut self, msg: String)-> Result<(), Error> {
+    validate_message(&msg)?;
+    self.commit_message(msg);
+    Ok(())
+}
+```
+You can [try this in the hax playground](https://hax-playground.cryspen.com/#fstar/4dec0f8ea9/gist=bdf9aa71a0c94a0318bc1682dd03d2a4) and see that F* is able to prove this post-condition!
+## Verifying sandwich
+In [collaboration](https://cryspen.com/post/hax-sandbox/) with SandboxAQ, we worked on verifying the [sandwich](https://github.com/sandbox-quantum/sandwich) library. Sandwich is a Rust high-level cryptographic library that wraps around lower-level libraries like OpenSSL. OpenSSL is a C library and sandwich interfaces with it using Rust/C bindings which means it was impossible for us to push verification to the lowest level because of raw pointers that are not supported by hax. Instead, we focused on more abstract properties about the OpenSSL API functions that need to be called (in the right order, with the right parameters) to make sure encryption is correctly set. We used the methodology presented above to prove these properties, see [this blogpost for more information](todo-link).
```

Comment:
I woulk cut "which means it was impossible .. not supported by hax". Instead, I would write, and we analyze the code starting from the public Sandwich API all the way down to just above the OpenSSL Rust bindings.